README.md
# URL2PNG plugin for Craft CMS 3.x

Lets you embed screenshots of webpages via the URL2PNG API.

![Screenshot](resources/img/icon.png)

## Requirements

This plugin requires Craft CMS 3.0.0-beta.23 or later.

## Installation

To install the plugin, follow these instructions.

1. Open your terminal and go to your Craft project:

        cd /path/to/project

2. Then tell Composer to load the plugin:

        composer require superbig/craft-url2png

3. In the Control Panel, go to Settings → Plugins and click the “Install” button for URL2PNG.

## Configuring URL2PNG

Place a copy of src/config.php in your Craft config directory, by default `config/`.

```php
<?php
return [
    'apiKey'    => '',
    'apiSecret' => '',
];
```

## Using URL2PNG

### Image embed
```twig
{{ craft.url2png.img({
            url: 'http://buildwithcraft.com',
            width: 500
}) }}
```

### URL
```twig
{% set img = craft.url2png.url({
        url: 'http://buildwithcraft.com',
        thumbnail_max_width: 500
        }) %}

<div style="background-image: url({{ img }}); width: 500px; height: 500px;"></div>
```

## Options

| Name                     | Description                                                          | Default  |
| ------------------------ | :------------------------------------------------------------------ | ------: |
| url                      | The url of the webpage you want to fetch                             ||
| thumbnail_max_width      | Constrain screenshot based on width. i.e. 500                        | 1:1 |
| viewport                 | Set viewport dimensions, adjust to your hearts content. i.e. 500x500 | 1480x1037 |
| fullpage                 | Will attempt to capture entire document canvas.                      | false |
| unique                   | Forces a fresh screenshot by varying this value. i.e. a timestamp.   ||
| user_agent               | Pass a custom user user agent                                        ||
| accept_languages         | Override the default HTTP Accept-Language header.                    | en-US,en;q=0.8 |
| custom_css_url           | Fetches a CSS stylesheet and injects it.                             ||
| say_cheese               | Delay screenshot until ```<div id='url2png-cheese'></div>``` is available. ||
| ttl                      | Set the TTL or "time to live" value for a screenshot in seconds. Defaults to 2592000 (30 days) ||
| alt                      | Sets the img alt attribute                                           ||
| class                    | Sets the img class attribute                                         ||
| width                    | Sets the img width attribute                                         ||
| height                   | Sets the img height attribute                                        ||

Brought to you by [Superbig](https://superbig.co)


CHANGELOG.md
# URL2PNG Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## 1.0.1 - 2018-05-29
### Fixed
- Fixed missing whitespace between src and img attributes

## 1.0.0 - 2018-05-25
### Added
- Initial release


src/URL2PNG.php
<?php
/**
 * URL2PNG plugin for Craft CMS 3.x
 *
 * Lets you embed screenshots of webpages via the URL2PNG API.
 *
 * @link      https://superbig.co
 * @copyright Copyright (c) 2018 Superbig
 */

namespace superbig\url2png;

use superbig\url2png\services\URL2PNGService as URL2PNGServiceService;
use superbig\url2png\variables\URL2PNGVariable;
use superbig\url2png\models\Settings;

use Craft;
use craft\base\Plugin;
use craft\services\Plugins;
use craft\events\PluginEvent;
use craft\web\UrlManager;
use craft\web\twig\variables\CraftVariable;
use craft\events\RegisterUrlRulesEvent;

use yii\base\Event;

/**
 * Class URL2PNG
 *
 * @author    Superbig
 * @package   URL2PNG
 * @since     1.0.0
 *
 * @property  URL2PNGServiceService $service
 * @method   Settings              getSettings()
 */
class URL2PNG extends Plugin
{
    // Static Properties
    // =========================================================================

    /**
     * @var URL2PNG
     */
    public static $plugin;

    // Public Properties
    // =========================================================================

    /**
     * @var string
     */
    public $schemaVersion = '1.0.0';

    // Public Methods
    // =========================================================================

    /**
     * @inheritdoc
     */
    public function init()
    {
        parent::init();
        self::$plugin = $this;

        Event::on(
            CraftVariable::class,
            CraftVariable::EVENT_INIT,
            function(Event $event) {
                /** @var CraftVariable $variable */
                $variable = $event->sender;
                $variable->set('url2png', URL2PNGVariable::class);
            }
        );

        Craft::info(
            Craft::t(
                'url2png',
                '{name} plugin loaded',
                ['name' => $this->name]
            ),
            __METHOD__
        );
    }

    // Protected Methods
    // =========================================================================

    /**
     * @inheritdoc
     */
    protected function createSettingsModel()
    {
        return new Settings();
    }

    /**
     * @inheritdoc
     */
    protected function settingsHtml(): string
    {
        return Craft::$app->view->renderTemplate(
            'url2png/settings',
            [
                'settings' => $this->getSettings(),
            ]
        );
    }
}


src/models/Settings.php
<?php
/**
 * URL2PNG plugin for Craft CMS 3.x
 *
 * Lets you embed screenshots of webpages via the URL2PNG API.
 *
 * @link      https://superbig.co
 * @copyright Copyright (c) 2018 Superbig
 */

namespace superbig\url2png\models;

use superbig\url2png\URL2PNG;

use Craft;
use craft\base\Model;

/**
 * @author    Superbig
 * @package   URL2PNG
 * @since     1.0.0
 */
class Settings extends Model
{
    // Public Properties
    // =========================================================================

    /**
     * @var string
     */
    public $apiKey = '';

    /**
     * @var string
     */
    public $apiSecret = '';

    // Public Methods
    // =========================================================================

    /**
     * @inheritdoc
     */
    public function rules()
    {
        return [
            [['apiKey', 'apiSecret'], 'string'],
            [['apiKey', 'apiSecret'], 'required'],
        ];
    }
}


src/models/URL2PNGModel.php
<?php
/**
 * URL2PNG plugin for Craft CMS 3.x
 *
 * Lets you embed screenshots of webpages via the URL2PNG API.
 *
 * @link      https://superbig.co
 * @copyright Copyright (c) 2018 Superbig
 */

namespace superbig\url2png\models;

use superbig\url2png\URL2PNG;

use Craft;
use craft\base\Model;

/**
 * @author    Superbig
 * @package   URL2PNG
 * @since     1.0.0
 */
class URL2PNGModel extends Model
{
    // Public Properties
    // =========================================================================

    /**
     * @var string
     */
    public $someAttribute = 'Some Default';

    // Public Methods
    // =========================================================================

    /**
     * @inheritdoc
     */
    public function rules()
    {
        return [
            ['someAttribute', 'string'],
            ['someAttribute', 'default', 'value' => 'Some Default'],
        ];
    }
}


src/variables/URL2PNGVariable.php
<?php
/**
 * URL2PNG plugin for Craft CMS 3.x
 *
 * Lets you embed screenshots of webpages via the URL2PNG API.
 *
 * @link      https://superbig.co
 * @copyright Copyright (c) 2018 Superbig
 */

namespace superbig\url2png\variables;

use craft\helpers\Template;
use superbig\url2png\URL2PNG;

use Craft;

/**
 * @author    Superbig
 * @package   URL2PNG
 * @since     1.0.0
 */
class URL2PNGVariable
{
    // Public Methods
    // =========================================================================

    public function img($settings = [])
    {
        $src        = $this->_run($settings);
        $attributes = $this->imgAttributes($settings);
        $html       = "<img src=\"{$src}\" {$attributes}>";

        return Template::raw($html);
    }

    public function url($settings = [])
    {
        return $this->_run($settings);
    }

    private function _run($settings)
    {
        return URL2PNG::$plugin->service->create($settings);
    }

    private function imgAttributes($settings)
    {
        return URL2PNG::$plugin->service->imgAttributes($settings);
    }
}


src/services/URL2PNGService.php
<?php
/**
 * URL2PNG plugin for Craft CMS 3.x
 *
 * Lets you embed screenshots of webpages via the URL2PNG API.
 *
 * @link      https://superbig.co
 * @copyright Copyright (c) 2018 Superbig
 */

namespace superbig\url2png\services;

use superbig\url2png\URL2PNG;

use Craft;
use craft\base\Component;
use yii\base\Exception;

/**
 * @author    Superbig
 * @package   URL2PNG
 * @since     1.0.0
 */
class URL2PNGService extends Component
{

    protected $settings;
    protected $allowedOptions;

    // Public Methods
    // =========================================================================

    public function init()
    {
        $this->settings = URL2PNG::$plugin->getSettings();

        // Set allowed options
        $this->setAllowedOptions();
    }

    public function create($options)
    {
        # Get your apikey from http://url2png.com/plans
        $apiKey    = $this->settings->apiKey;
        $apiSecret = $this->settings->apiSecret;

        // Clean options
        $options = $this->cleanPassedOptions($options, $this->allowedOptions);

        if (!is_array($options) && in_array('url', $options)) {
            throw new Exception(Craft::t('url2png', 'url2png: No url defined'));
        }

        # create the query string based on the options
        foreach ($options as $key => $value) {
            $_parts[] = "$key=$value";
        }

        # create a token from the ENTIRE query string
        $query_string = implode("&", $_parts);
        $token        = md5($query_string . $apiSecret);

        return "https://api.url2png.com/v6/{$apiKey}/{$token}/png/?{$query_string}";
    }

    private function setAllowedOptions()
    {
        $this->allowedOptions = [
            'url',
            'thumbnail_max_width', // Default 1:1, can be something like 500
            'viewport', // Defaults to 1480x1037
            'fullpage', // Defaults to false
            'unique', // Forces a fresh screenshot by sending a unique value. A timestamp is suggested.
            'user_agent', // Custom user agent
            'accept_languages', // Accept-Language header. Defaults to en-US,en;q=0.8
            'custom_css_url', // Fetches a CSS stylesheet and injects it
            'say_cheese', // Delay screenshot until <div id='url2png-cheese'></div> is available.
            'ttl', // Set the TTL or "time to live" value for a screenshot in seconds. Defaults to 2592000 (30 days)
        ];
    }

    private function cleanPassedOptions($options, $allowedOptions)
    {
        $cleanOptions = [];

        foreach ($options as $key => $value) {
            // Check if this option is allowed
            if (!in_array($key, $allowedOptions)) {
                Craft::error(
                    Craft::t('url2png', 'Option {key} not allowed', ['key' => $key]),
                    'url2png'
                );

                continue;
            }

            // If this is url, urlencode it
            if ($key === 'url') $value = urlencode($value);

            $cleanOptions[ $key ] = $value;
        }

        return $cleanOptions;
    }

    public function imgAttributes($options = [])
    {
        $allowedAttributes = [
            'class',
            'width',
            'height',
            'alt',
        ];
        $attributes        = $this->cleanPassedOptions($options, $allowedAttributes);

        if (empty($attributes)) {
            return '';
        }

        $attributes = array_map(function($key, $value) {
            return "{$key}=\"{$value}\"";
        }, array_keys($attributes), $attributes);

        return implode(' ', $attributes);
    }
}
